### DISCUSSION TITLE
Color / theme-related bugs

### DISCUSSION THREAD

**Discussion by djmitche** [GothenburgBitFactory/taskwarrior#3692] [General]
I'm glad to see @d2718nis filing some really well-written issues regarding terminal colors / themes.

There's actually [a bunch of those issues in the topic:themes label](https://github.com/GothenburgBitFactory/taskwarrior/issues?q=is%3Aissue+is%3Aopen+label%3Atopic%3Athemes). Maybe there are some more comprehensive changes to be made? For example, I see that a few of those relate to the calendar view -- perhaps there's an underlying bug causing all of those?

At any rate, I just wanted to celebrate the increased attention to this part of Taskwarrior!

**Comment by d2718nis** [comment]
I'm glad my input has been helpful! I'm currently working on a [Rosé Pine](https://rosepinetheme.com/) inspired theme, which is why I'm noticing some of these details. At one point, I considered asking if you'd be interested in adding it to the repo, but it's still a work in progress.

Although C++ isn't my primary tool of choice, I'm comfortable working with it enough to be able to help with fixing things here and there. Regarding the `calendar` report, I wouldn't say there's a bug per se, but it seems like it's [not utilizing](https://github.com/GothenburgBitFactory/taskwarrior/blame/develop/src/commands/CmdCalendar.cpp#L268-L273) the [Color Rules](https://github.com/GothenburgBitFactory/taskwarrior/blob/develop/src/rules.cpp). It could benefit from a bit of refactoring to take advantage of that dynamic approach.

**Reply by djmitche** [reply]
Regarding themes, that sounds fine to me! In wonder if we should think forward to some kind of showcase of themes, rather than just including a whole bunch with the distribution? The [current list](https://taskwarrior.org/docs/themes/) is a bit out of date and probably doesn't represent individual users' terminals very well (with differences in fonts, etc.). Maybe there's a way to demo themes on the command line?

Regarding calendar, that sounds like a good fix! Despite my growing list of commits to Chrome, I stand by the assertion that C++ is also not my primary tool of choice, but we all muddle along -- it's the C++ way.

**Reply by d2718nis** [reply]
> if we should think forward to some kind of showcase of themes

There's [xterm.js](https://github.com/xtermjs/xterm.js/), which emulates a terminal environment in web/electron apps or even pipes to bash. Theoretically, it's possible to create a web app showcasing themes with adjustable settings including fonts. That said, it might not justify the time investment.

> Maybe there's a way to demo themes on the command line?

This would be doable if the `rc.include=` override worked to include an entire theme file, but it doesn't (related [issue](https://github.com/GothenburgBitFactory/taskwarrior/issues/1552), wontfix). Instead, it’s possible to load color settings one by one from the file, but I believe would require writing some logic. For now I use `include $TW_THEME` in my `taskrc` for dynamic theme switching while testing: the theme I’m building includes three variants — or six if you count both `-16` and `-256` color themes.

I’ve also considered building a standard testing tool, allowing consistent screenshots for themes in the repo `README` (useful if colors change later). It's still up in the air, though.

Regarding a separate *theme context* command, I feel it's easy enough to edit one line in the configuration file and test reports. However, a command like `task showcase solarized-dark-256` that outputs a set of default reports using a given theme could be an elegant solution. Did you have something similar in mind, or other ideas?

> Regarding calendar, that sounds like a good fix!

I'll see what I can do! There are a couple of other issues on my list as well, so I’ll add it to the mix.

**Reply by djmitche** [reply]
I didn't have anything in particular in mind. `task showcase` sounds cool, as does the addition of `xterm.js` in the TW docs. A bug for either or both of those would be great! (`tw.org` is the repo for the Taskwarrior docs)

**Reply by d2718nis** [reply]
I’ll see what I can do! I’ll definitely reach out here or in the related issues with any questions to make sure my ideas align with the project’s direction. Thanks!
